SemanticUI/Modules/Dropdown.elm

```diff
@@ -0,0 +1,438 @@
+                    , Dropdown.item div [] [ text "Dropbox" ]
+                    , Dropdown.item div [] [ text "Adobe Creative Cloud" ]
+                    , Dropdown.item div [] [ text "Private FTP" ]
+                    , Dropdown.item div [] [ text "Another Service..." ]
+                    ]
+                ]
+            ]
+        ]
+
+-}
+
+import Dropdown
+import Html exposing (..)
+import Html.Attributes exposing (..)
+import Html.Events exposing (..)
+import Json.Decode as Decode
+import SemanticUI.Modules.Common exposing (HtmlBuilder)
+
+
+{-| Everything needed to build a particular dropdown.
+
+A function that takes a record with the following functions
+
+  - toDropdown - the function `root` with `Config` and `State` applied to it.
+  - toToggle - the function `toggle` with `Config` and `State` applied to it.
+  - toDrawer - the function `drawer` with `Config` and `State` applied to it.
+  - toItem - the function `item` with `Config` and `State` applied to it.
+
+The final resultant value is what you decide (some kind of DOM)
+
+-}
+type alias DropdownBuilder msg =
+    { toDropdown : HtmlBuilder msg -> HtmlBuilder msg
+    , toToggle : HtmlBuilder msg -> HtmlBuilder msg
+    , toDrawer : HtmlBuilder msg -> HtmlBuilder msg
+    , toItem : HtmlBuilder msg -> HtmlBuilder msg
+    }
+
+
+{-| Define when the dropdown should open
+-}
+type ToggleEvent
+    = OnClick
+    | OnHover
+    | OnFocus
+
+
+{-| Configuration for a dropdown
+
+  - identifier - Unique identifier for the dropdown
+  - onToggle - Handle state change of the dropdown
+  - toggleEvent - When should the dropdown expand (default OnClick)
+  - readOnly - Is the dropdown read only (default False)
+
+-}
+type alias Config msg =
+    { identifier : String
+    , onToggle : State -> msg
+    , toggleEvent : ToggleEvent
+    , readOnly : Bool
+    }
+
+
+{-| Create a default Config for a dropdown.
+
+Takes:
+
+  - The unique identifier.
+  - The state change handler.
+
+-}
+init : { config | identifier : String, onToggle : State -> msg } -> Config msg
+init { identifier, onToggle } =
+    { identifier = identifier
+    , onToggle = onToggle
+    , toggleEvent = OnClick
+    , readOnly = False
+    }
+
+
+{-| Set toggleEvent on a Config
+-}
+toggleEvent : ToggleEvent -> { config | toggleEvent : ToggleEvent } -> { config | toggleEvent : ToggleEvent }
+toggleEvent a config =
+    { config | toggleEvent = a }
+
+
+{-| Set readOnly on a Config
+-}
+readOnly : Bool -> { config | readOnly : Bool } -> { config | readOnly : Bool }
+readOnly a config =
+    { config | readOnly = a }
+
+
+{-| The current state of the dropdown drawer
+-}
+type State
+    = Closed
+    | Opening
+    | Opened
+    | Closing
+
+
+{-| Render a dropdown.
+
+Takes:
+
+  - Configuration
+  - The current state of the dropdown
+  - The layout function for the dropdown
+
+-}
+dropdown :
+    Config msg
+    -> State
+    -> (DropdownBuilder msg -> html)
```

Comment:
Could this be part of the `Config` for a `dropdown`?

Reply:
It's a little awkward, but yes! Would need to be a slightly different type than for config given to `root` / `drawer` / `toggle` etc in that case.

I suppose I could go with

```elm
type alias Config config = {
  ...
}

type alias LayoutConfig = Config { layout : DropdownBuilder msg -> html }
```

if that seems pleasant to use?

Reply:
It should be fine with just a single `Config` type, parameterised on `msg` and `html`.

Reply:
I'm not sure if I follow? Currently the following top-level functions are exposed by the module:

```
root : Config msg -> State -> HtmlBuilder msg -> HtmlBuilder msg
drawer : Config msg -> State -> HtmlBuilder msg -> HtmlBuilder msg
toggle : Config msg -> State -> HtmlBuilder msg -> HtmlBuilder msg
```

where `Config msg` does not include the layout.

I'm not sure whether these ought to be exposed, but that's the situation as it stands.

Reply:
This is probably no longer relevant, but sparked a bunch of follow up work.